[Linux] Fix event loop task consecutive start/stop

arkq opened the pull request:
### Problem

On Linux (POSIX implementation) it's not possible to start matter event loop task after calling `PlatformMgr().StopEventLoopTask()`. Also, it's not possible to reinitialize matter stack, because in that case event loop will not run as well.

### Changes

- reset `mShouldRunEventLoop` flag on `_InitChipStack()` and `_StartEventLoopTask()`
- modified TestPlatformMgr to properly cover that use case scenario

### Testing

CI will verify.

andy31415 commented on src/platform/tests/TestPlatformMgr.cpp:
```diff
@@ -55,14 +56,29 @@ static void TestPlatformMgr_InitShutdown(nlTestSuite * inSuite, void * inContext
+        // scheduling few lambdas (for the test to pass, the event loop will
+        // have to process more than one event).
+        DeviceLayer::SystemLayer().ScheduleLambda([&counterRun]() { counterRun++; });
+        chip::test_utils::SleepMillis(10);
```
Can we avoid sleep in tests? Could we use some cond/barriers?

arkq replied:
To be honest I don't know how to implement that without a sleep (somewhere). The problem is that it seems (I have not analyzed the code in details) that `LayerImplSelect::HandleEvents()` handles all pending events at once. Without the sleep between schedule calls (I can implement that with a timer, but technically it will be a sleep anyway) I've got only one iteration of `do { ... } while()` in `_RunEventLoop()`...

Also, one more thing is not clear for me. Maybe someone will explain that. Why we've got `do {...} while` instead of `while(){...}`? If that loop would be implemented in the latter form, the bug would have been detected earlier, because this test runner starts and stops event loop task in few test cases and it works only because event loop always executes `HandleEvents()`.

andy31415 replied:
Lets add this explanation at the top of the sleep, to guide others to remove it if applicable.

andy31415 commented on src/platform/tests/TestPlatformMgr.cpp:
```diff
@@ -55,14 +56,29 @@ static void TestPlatformMgr_InitShutdown(nlTestSuite * inSuite, void * inContext
 
-    err = PlatformMgr().StopEventLoopTask();
-    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counterRun == (3 * 2));
```
We have `schedulelambda` followed immediately by `stop`. There seems to be no guarantee that the lambda would run unless we barrier things.

arkq replied:
I've added a simple barrier for syncing schedule calls.

andy31415 commented on src/platform/tests/TestPlatformMgr.cpp:
```diff
@@ -55,14 +56,29 @@ static void TestPlatformMgr_InitShutdown(nlTestSuite * inSuite, void * inContext
 
-    err = PlatformMgr().StopEventLoopTask();
-    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counterRun == (3 * 2));
```
We have `schedulelambda` followed immediately by `stop`. There seems to be no guarantee that the lambda would run unless we barrier things.

andy31415 replied:
We have `schedulelambda` followed immediately by `stop`. There seems to be no guarantee that the lambda would run unless we barrier things.

andy31415 commented:
@arkq it seems we need some fixes regarding std::mutex.

arkq commented:
> @arkq it seems we need some fixes regarding std::mutex.

I've changed sync implementation. Open IoT SDK seems to be a platform for MCUs which might use RTOS, so using mutex/condition_variables will be OS-specific. I've changed that sync to use simple busy loop with timeout, so it will work on all platforms.
